Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp

```diff
@@ -1244,15 +1244,38 @@ void WebLocalFrameLoaderClient::updateGlobalHistoryRedirectLinks()
     }
 }
 
-bool WebLocalFrameLoaderClient::shouldGoToHistoryItem(HistoryItem& item) const
+bool WebLocalFrameLoaderClient::shouldGoToHistoryItem(HistoryItem& item, bool sameDocumentNavigation) const
 {
+    // In WebKit2, the synchronous version of this policy client should only ever be consulted for same document navigations.
+    RELEASE_ASSERT(sameDocumentNavigation);
+
     RefPtr webPage = m_frame->page();
     if (!webPage)
         return false;
-    webPage->send(Messages::WebPageProxy::WillGoToBackForwardListItem(item.itemID(), item.isInBackForwardCache()));
+
+    auto sendSyncResult = webPage->sendSync(Messages::WebPageProxy::ShouldGoToBackForwardListItemSync(item.itemID()));
+    auto [shouldGo] = sendSyncResult.takeReplyOr(true);
+    return shouldGo;
+}
+
+bool WebLocalFrameLoaderClient::supportsAsyncShouldGoToHistoryItem() const
+{
     return true;
 }
 
+void WebLocalFrameLoaderClient::shouldGoToHistoryItemAsync(HistoryItem& item, CompletionHandler<void(bool)>&& completionHandler) const
+{
+    RefPtr webPage = m_frame->page();
+    if (!webPage) {
+        completionHandler(false);
+        return;
+    }
+
+    webPage->sendWithAsyncReply(Messages::WebPageProxy::ShouldGoToBackForwardListItem(item.itemID(), item.isInBackForwardCache()), [completionHandler = WTFMove(completionHandler)] (bool result) mutable {
```

Comment:
Would this work?
```
webPage->sendWithAsyncReply(Messages::WebPageProxy::ShouldGoToBackForwardListItem(item.itemID(), item.isInBackForwardCache()), WTFMove(completionHandler));
```

Reply:
Exact same comment as above. Extra lambda no longer needed.